File: src/main/java/pers/neige/neigeitems/action/node/impl/RepeatParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import lombok.var;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.script.CompiledScript;
import pers.neige.neigeitems.utils.ListUtils;
import pers.neige.neigeitems.utils.NumberParser;
import pers.neige.neigeitems.utils.StringUtils;

import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

/**
 * 对齐旧 section RepeatParser：支持 transform；非法/缺省 repeat 默认 1；无上限。
 */
public class RepeatParser extends NodeParser {
    /**
     * 用于 repeat 节点的已编译 transform 脚本缓存（与旧 section 一致）。
     */
    private static final ConcurrentHashMap<String, CompiledScript> compiledScripts = new ConcurrentHashMap<>();

    public RepeatParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "repeat";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            context,
            getParsedValue(context, params, "content"),
            getParsedValue(context, params, "repeat"),
            getParsedValue(context, params, "separator"),
            getParsedValue(context, params, "prefix"),
            getParsedValue(context, params, "postfix"),
            // transform 文本本身不做节点解析，对齐旧 section
            params.getString("transform")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\', 3);
        if (paramsList.isEmpty()) {
            warning("缺少 content 参数");
            return null;
        }
        return handle(
            context,
            paramsList.get(0),
            ListUtils.getOrNull(paramsList, 1),
            ListUtils.getOrNull(paramsList, 2),
            null,
            null,
            null
        );
    }

    private @Nullable String handle(
        @NonNull ActionContext context,
        @Nullable String content,
        @Nullable String repeatText,
        @Nullable String separator,
        @Nullable String prefix,
        @Nullable String postfix,
        @Nullable String rawTransform
    ) {
        // 旧：content 默认 ""
        val realContent = content == null ? "" : content;
        // 旧：toIntOrNull 失败 → 默认 1；coerceAtLeast(0)
        int length = 1;
        if (repeatText != null && !repeatText.isEmpty()) {
            val parsed = NumberParser.parseInteger(repeatText);
            if (parsed != null) {
                length = Math.max(parsed, 0);
            }
        }

        CompiledScript transform = null;
        if (rawTransform != null && !rawTransform.isEmpty()) {
            transform = compiledScripts.computeIfAbsent(rawTransform, script -> new CompiledScript(
                "function main() {\n" +
                    script + "\n" +
                    "}"
            ));
        }

        val result = new StringBuilder();
        if (prefix != null) {
            result.append(prefix);
        }

        val map = new HashMap<String, Object>();
        if (transform != null) {
            val caster = context.getCaster();
            if (caster instanceof OfflinePlayer) {
                map.put("player", caster);
            }
            map.put("it", realContent);
            map.put("vars", (Function<String, String>) string -> this.manager.parseNode(string, context));
        }

        for (int index = 0; index < length; index++) {
            var element = realContent;
            if (transform != null) {
                map.put("index", index);
                val transformed = transform.invoke("main", map);
                element = transformed == null ? "" : transformed.toString();
            }
            result.append(element);
            if (separator != null && index != length - 1) {
                result.append(separator);
            }
        }
        if (postfix != null) {
            result.append(postfix);
        }
        return result.toString();
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/RepeatParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.script.CompiledScript
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.util.concurrent.ConcurrentHashMap

/**
 * repeat节点解析器
 */
object RepeatParser : SectionParser() {
    override val id: String = "repeat"

    /**
     * 获取所有用于repeat节点的已编译的js脚本文件及文本
     */
    val compiledScripts = ConcurrentHashMap<String, CompiledScript>()

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            data.getString("content")?.parseSection(cache, player, sections) ?: "",
            data.getString("separator")?.parseSection(cache, player, sections),
            data.getString("prefix")?.parseSection(cache, player, sections),
            data.getString("postfix")?.parseSection(cache, player, sections),
            data.getString("repeat")?.parseSection(cache, player, sections)?.toIntOrNull(),
            data.getString("transform")
        )
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param content 待重复内容
     * @param separator 分隔符
     * @param prefix 前缀
     * @param postfix 后缀
     * @param repeat 长度限制
     * @param transform 操作函数
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        content: String,
        separator: String?,
        prefix: String?,
        postfix: String?,
        repeat: Int?,
        transformString: String?,
    ): String {
        // 获取长度限制
        val length = (repeat ?: 1).coerceAtLeast(0)
        // 获取操作函数
        val transform = transformString?.let {
            compiledScripts.computeIfAbsent(it) {
                CompiledScript(
                    """
                    function main() {
                        $it
                    }""".trimIndent()
                )
            }
        }

        // 开始构建结果
        val result = StringBuilder()
        // 添加前缀
        prefix?.let {
            result.append(it)
        }

        // 预定义参数map
        val map = HashMap<String, Any>()
        // 预添加参数
        transform?.let {
            player?.let {
                // 玩家
                map["player"] = player
            }
            // 待操作元素
            map["it"] = content
            // 节点解析函数
            map["vars"] =
                java.util.function.Function<String, String> { string -> string.parseSection(cache, player, sections) }
        }

        for (index in 0 until length) {
            // 解析元素节点
            var element = content
            // 操作元素
            transform?.let {
                // 当前序号
                map["index"] = index
                // 操作元素
                element = transform.invoke("main", map)?.toString() ?: ""
            }
            // 添加元素
            result.append(element)
            // 添加分隔符
            separator?.let {
                if (index != (length - 1)) {
                    result.append(it)
                }
            }
        }
        // 添加后缀
        postfix?.let {
            result.append(it)
        }
        // 返回结果
        return result.toString()
    }
}